File: MKAgent/Protocol.java
package MKAgent;
/**
 * Creates messages to be sent and interprets messages received.
 */
public class Protocol
{
	/**
	 * An object of this type is returned by interpretStateMsg().
	 * @see Protocol#interpretStateMsg(String, Board)
	 */
	public static class MoveTurn
	{
		/**
		 * "true" if the game is over, "false" otherwise.
		 */
		public boolean end;
		/**
		 * "true" if it's this agent's turn again, "false" otherwise.
		 */
		public boolean again;
		/**
		 * The number of the hole that characterises the move which has been
		 * made (the move starts with picking the seeds from the given hole)
		 * or -1 if the opponent has made a swap.
		 */
		public int move;
	}


    /**
     * Creates a "move" message.
     * @param hole The hole to pick the seeds from.
	 * @return The message as a string.
     */
    public static String createMoveMsg (int hole)
    {
		return "MOVE;" + hole + "\n";
    }

    /**
     * Creates a "swap" message.
	 * @return The message as a string.
     */
    public static String createSwapMsg ()
    {
		return "SWAP\n";
    }

    /**
     * Determines the type of a message received from the game engine. This
     * method does not check whether the message is well-formed.
     * @param msg The message.
     * @return The message type.
     * @throws InvalidMessageException if the message type cannot be
     *         determined.
     */
    public static MsgType getMessageType (String msg) throws InvalidMessageException
    {
    	if (msg.startsWith("START;"))
    		return MsgType.START;
    	else if (msg.startsWith("CHANGE;"))
    		return MsgType.STATE;
    	else if (msg.equals("END\n"))
    		return MsgType.END;
    	else
    		throw new InvalidMessageException("Could not determine message type.");
    }

    /**
     * Interprets a "new_match" message. Should be called if
     * getMessageType(msg) returns MsgType.START
     * @param msg The message.
     * @return "true" if this agent is the starting player (South), "false"
     *         otherwise.
     * @throws InvalidMessageException if the message is not well-formed.
     * @see #getMessageType(String)
     */
    public static boolean interpretStartMsg (String msg) throws InvalidMessageException
    {
    	if (msg.charAt(msg.length()-1) != '\n')
    		throw new InvalidMessageException("Message not terminated with 0x0A character.");

    	String position = msg.substring(6, msg.length()-1);
    	if (position.equals("South"))
    		return true;  // South starts the game
    	else if (position.equals("North"))
    		return false;
    	else
    		throw new InvalidMessageException("Illegal position parameter: " + position);
    }

    /**
     * Interprets a "state_change" message. Should be called if
     * getMessageType(msg) returns MsgType.STATE 
     * @param msg The message.
     * @param board This is an output parameter. It will store the new state
     *        of the Kalah board. The board has to have the right dimensions
     *        (number of holes), otherwise an InvalidMessageException is
     *        thrown.
     * @return information about the move that led to the state change and
     *         who's turn it is next.
     * @throws InvalidMessageException if the message is not well-formed.
     * @see #getMessageType(String)
     */
    public static MoveTurn interpretStateMsg (String msg, Board board) throws InvalidMessageException
    {
    	MoveTurn moveTurn = new MoveTurn();

    	if (msg.charAt(msg.length()-1) != '\n')
    		throw new InvalidMessageException("Message not terminated with 0x0A character.");

    	String[] msgParts = msg.split(";", 4);
    	if (msgParts.length != 4)
    		throw new InvalidMessageException("Missing arguments.");

    	// msgParts[0] is "CHANGE"

    	// 1st argument: the move (or swap)
    	if (msgParts[1].equals("SWAP"))
    		moveTurn.move = -1;
    	else
    	{
        	try
        	{
        		moveTurn.move = Integer.parseInt(msgParts[1]);
        	}
        	catch (NumberFormatException e)
        	{
        		throw new InvalidMessageException("Illegal value for move parameter: " + e.getMessage());
        	}
    	}

    	// 2nd argument: the board
    	String[] boardParts = msgParts[2].split(",", -1);
    	/*if (boardParts.length % 2 != 0)
    		throw new InvalidMessageException("Malformed board: odd number of entries.");*/
    	if (2*(board.getNoOfHoles()+1) != boardParts.length)
    		throw new InvalidMessageException("Board dimensions in message ("
    				+ boardParts.length + " entries) are not as expected ("
    				+ 2*(board.getNoOfHoles()+1) + " entries).");
    	try
    	{
    		// holes on the north side:
    		for (int i = 0; i < board.getNoOfHoles(); i++)
    			board.setSeeds(Side.NORTH, i+1, Integer.parseInt(boardParts[i]));
    		// northern store:
			board.setSeedsInStore(Side.NORTH, Integer.parseInt(boardParts[board.getNoOfHoles()]));
			// holes on the south side:
       		for (int i = 0; i < board.getNoOfHoles(); i++)
    			board.setSeeds(Side.SOUTH, i+1, Integer.parseInt(boardParts[i+board.getNoOfHoles()+1]));
    		// southern store:
			board.setSeedsInStore(Side.SOUTH, Integer.parseInt(boardParts[2*board.getNoOfHoles()+1]));
    	}
    	catch (NumberFormatException e)
    	{
    		throw new InvalidMessageException("Illegal value for seed count: " + e.getMessage());
    	}
    	catch (IllegalArgumentException e)
    	{
    		throw new InvalidMessageException("Illegal value for seed count: " + e.getMessage());
    	}

    	// 3rd argument: who's turn?
		moveTurn.end = false;
    	if (msgParts[3].equals("YOU\n"))
    		moveTurn.again = true;
    	else if (msgParts[3].equals("OPP\n"))
    		moveTurn.again = false;
    	else if (msgParts[3].equals("END\n"))
    	{
    		moveTurn.end = true;
    		moveTurn.again = false;
    	}
    	else
    		throw new InvalidMessageException("Illegal value for turn parameter: " + msgParts[3]);

    	return moveTurn;
    }
}

File: magent/protocol/protocol.py
import sys
import logging
from magent.protocol.msg_type import MsgType
from magent.protocol.move_turn import MoveTurn
from magent.protocol.invalid_message_exception import InvalidMessageException


def send_msg(msg: str):
    """
        Send a message to the game engine.
        @:param msg the message
    """
    logging.info("Message to be sent: " + msg)
    print(msg, end="\n")
    sys.stdout.flush()


def read_msg() -> str:
    """
        Receives a message from the game engine. Messages are terminated by
        a '\n' character.

        @:return The message.
        @:raise IOException if there has been an I/O error.
    """

    msg = sys.stdin.readline()
    logging.info('Received: ' + msg)
    return msg


def create_move_msg(hole: int) -> str:
    """
        Creates a move message.
        @:param hole The hole to pick the seeds from.
        @:return The message as a string.
    """
    return "MOVE;" + str(hole)


def create_swap_msg() -> str:
    """
        Create a swap message.

        @:returnThe message as a string
    """
    logging.info("We swapped")
    return "SWAP"


def get_msg_type(msg: str) -> MsgType:
    if msg.startswith("START;"):
        logging.debug("Game engine sent a START command.")
        return MsgType.START
    elif msg.startswith("CHANGE;"):
        logging.debug("Game engine sent a CHANGE command.")
        return MsgType.STATE
    elif msg.startswith("END\n"):
        logging.debug("Game engine sent an END command.")
        return MsgType.END
    else:
        raise InvalidMessageException("Could not determine message type.")


def interpret_start_msg(msg: str) -> bool:
    """
        Interprets a "new_match" message. Should be called if
        getMessageType(msg) returns MsgType.START
        @:param msg The message.
        @:return "true" if this agent is the starting player (South), "false" otherwise.
        @:raises InvalidMessageException if the message is not well-formed.
    """
    if msg[-1] != '\n':
        raise InvalidMessageException("Message not terminated with 0x0A character.")

    # Message are of the form START:<POSITION> \n
    position = msg[6:-1]
    if position == "South":
        return True
    elif position == "North":
        return False
    else:
        raise InvalidMessageException("Illegal position parameter: " + position)


def interpret_state_msg(msg: str) -> MoveTurn:
    """
        Interprets a "state_change" message. Should be called if
        getMessageType(msg) returns MsgType.STATE

        @:param msg   The message.
        @:return information about the move that led to the state change and
        who's turn it is next.
        @:raises InvalidMessageException if the message is not well-formed.
    """
    move_turn = MoveTurn()
    if msg[-1] != '\n':
        raise InvalidMessageException("Message not terminated with 0x0A character.")

    msg_parts = msg.split(';', 4)

    if len(msg_parts) != 4:
        raise InvalidMessageException("Missing arguments.")
    # msgParts[0] is "CHANGE"
    # 1st argument: the move (or swap)

    if msg_parts[1] == "SWAP":
        logging.info("Opponent requested swap")
        move_turn.move = 0

    else:
        try:
            move_turn.move = int(msg_parts[1])
        except ValueError as e:
            raise InvalidMessageException("Illegal value for move parameter: ", str(e))

    # 3rd argument: who's turn
    move_turn.end = False
    if msg_parts[3] == "YOU\n":
        move_turn.again = True
    elif msg_parts[3] == "OPP\n":
        move_turn.again = False
    elif msg_parts[3] == "END\n":
        move_turn.end = True
        move_turn.again = False
    else:
        raise InvalidMessageException("Illegal value for turn parameter: " + msg_parts[3])

    logging.info("This was the move: " + str(move_turn.move))
    logging.info("Is the game over? " + str(move_turn.end))
    logging.info("Is it our turn again? " + str(move_turn.again))

    return move_turn
